snuba/datasets/plans/partitioned_storage.py

```diff
@@ -0,0 +1,153 @@
+
+import sentry_sdk
+
+from snuba.clickhouse.query_dsl.accessors import get_object_ids_in_query_ast
+from snuba.clickhouse.translators.snuba.mapping import TranslationMappers
+from snuba.clusters.cluster import ClickhouseCluster, get_cluster
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.datasets.partitioning import (
+    is_storage_set_partitioned,
+    map_logical_partition_to_physical_partition,
+    map_org_id_to_logical_partition,
+)
+from snuba.datasets.plans.query_plan import (
+    ClickhouseQueryPlan,
+    ClickhouseQueryPlanBuilder,
+)
+from snuba.datasets.plans.single_storage import (
+    SimpleQueryPlanExecutionStrategy,
+    get_query_data_source,
+)
+from snuba.datasets.plans.translator.query import QueryTranslator
+from snuba.datasets.storage import ReadableStorage
+from snuba.query.logical import Query as LogicalQuery
+from snuba.query.processors.physical import ClickhouseQueryProcessor
+from snuba.query.processors.physical.conditions_enforcer import (
+    MandatoryConditionEnforcer,
+)
+from snuba.query.processors.physical.mandatory_condition_applier import (
+    MandatoryConditionApplier,
+)
+from snuba.query.query_settings import QuerySettings
+from snuba.util import with_span
+
+
+class StorageClusterSelector(ABC):
+    """
+    The component provided by a dataset and used at the beginning of the
+    execution of a query to pick the storage set a query should be executed
+    onto.
+    """
+
+    @abstractmethod
+    def select_cluster(
+        self, query: LogicalQuery, query_settings: QuerySettings
+    ) -> ClickhouseCluster:
+        raise NotImplementedError
+
+
+class ColumnBasedStoragePartitionSelector(StorageClusterSelector):
+    """
+    Storage partition selector for the generic metrics storage. This is needed
+    because the generic metrics storage can be partitioned and we would need to
+    know which partition to use for a specific query.
+    """
+
+    def __init__(
+        self, storage_set: StorageSetKey, partition_key_column_name: str
+    ) -> None:
+        self.storage_set = storage_set
+        self.partition_key_column_name = partition_key_column_name
+
+    def select_cluster(
+        self, query: LogicalQuery, query_settings: QuerySettings
+    ) -> ClickhouseCluster:
+        """
+        Selects the cluster to use for a query if the storage set is partitioned.
+        If the storage set is not partitioned, it returns the default cluster.
+        """
+        if not is_storage_set_partitioned(self.storage_set):
+            return get_cluster(self.storage_set)
+
+        org_ids = get_object_ids_in_query_ast(query, self.partition_key_column_name)
+        assert org_ids is not None
+        assert len(org_ids) == 1
+        org_id = org_ids.pop()
+
+        physical_partition = map_logical_partition_to_physical_partition(
+            map_org_id_to_logical_partition(org_id)
+        )
+        cluster = get_cluster(self.storage_set, physical_partition)
```

Comment:
@onewland - When a query needs to span across clickhouse clusters, we need a way to find that out. Now if we are going to build a different storage set for the cross clickhouse cluster, things might start to get messy over here. Query comes for a dataset A. Dataset A is bound to storage set A. Processing decides it has to go across clusters. Change Storage set to B. That means modify ColumnBasedStoragePartitionSelector to take 2 storage sets. Which would mean when defining the dataset entities there would most likely be circluar imports. 

My suggestion is to have a config which can say whether to go across clusters or not but use the same storage set.

Reply:
Thinking about this a bit more, it might be ok. But this is something to be aware of

Reply:
I was sort of assuming that this call to `cluster = get_cluster(storage_set, slice_id)` would be changed when we introduced the run-time flag to

```
if override_cluster_exists(storage_set, logical_partition):
   cluster = get_cluster(get_override_cluster_storage_set_key(storage_set, slice_id))
else:
   cluster = get_cluster(storage_set, slice_id)
```

does that seem like a bad idea to you for some reason?